Add sphere-engine package

Eggbertx opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [x] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

danger-public commented on types/sphere-engine-browser/sphere1-api.d.ts:
```diff
@@ -0,0 +1,1380 @@
+ * @example
+ * let start = GetTime();
+ * while (GetTime() < start + 1000) {
+ * 	// do stuff
```
<!--
  1 failure:  The root .editorc...
  0 warning: 
  
  
  DangerID: danger-id-Danger;
  File: types/sphere-engine-browser/sphere1-api.d.ts;
  Line: 575;
-->
- :no_entry_sign: The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.

danger-public commented:
<!--
  1 failure:  The root .editorc...
  0 warning: 
  
  
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **types/sphere-engine-browser/sphere2-api.d.ts#L2113** - The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 9ffc35e611ae9a28072507323a992b44bd8a8adf
</p>

andrewbranch commented on types/sphere-engine-browser/index.d.ts:
```diff
@@ -0,0 +1,11 @@
+// Type definitions for the non-npm package sphere-engine-browser 5.5
+// Project: https://github.com/fatcerberus/miniSphere
+// Definitions by: Fat Cerberus <https://github.com/fatcerberus>, Eggbertx <https://github.com/eggbertx>
```
Did @fatcerberus work on these definitions with you, or just the source library? (Also their GitHub display name is Bruce Pascoe, not Fat Cerberus 😄)

Eggbertx replied:
He created the miniSphere engine itself and sphere2-api.d.ts (with my modifications to make it dtlint compatible), and I created sphere1-api.d.ts based on the original Sphere engine's API, which miniSphere is mostly backwards compatible with. And I put "Fat Cerberus" as the name in index.d.ts since that's what he put in sphere2-api.d.ts

andrewbranch replied:
Wait, what are sphere2-api.d.ts and sphere1-api.d.ts

fatcerberus commented:
miniSphere: https://github.com/fatcerberus/miniSphere
Oozaru: https://github.com/fatcerberus/oozaru

The definitions here are for the above. They are JavaScript-based game engines, miniSphere being the standalone version and Oozaru being the HTML5-based incarnation. The engine was initially based on the **Sphere** game engine written by Chad Austin, but was written from the ground up and has since evolved far past its original inspiration.

Getting down to brass tacks, there are two API versions:
* **Sphere v1**, which is the API supported by the original Sphere engine. This API is supported by miniSphere but not Oozaru (as it's not compatible with the browser event loop), and is covered by `sphere1-api.d.ts`.
* **Sphere v2**, the brand-new API introduced in miniSphere and also supported in Oozaru. This is the API intended to be used in new projects and is covered by `sphere2-api.d.ts`.

andrewbranch commented:
Does it make sense to triple-slash reference both versions of the API from the index file, then?

fatcerberus commented:
I'm not sure it does, since Oozaru doesn't support the v1 API at all. Only miniSphere supports both. Though I'm not that familiar with DT to know what standard procedure here is, so...

andrewbranch commented:
There’s a fair amount of flexibility here since this doesn’t need to exactly mirror a versioned npm package, but it seems like it would be reasonable to use [versioned directories](https://github.com/DefinitelyTyped/DefinitelyTyped#if-a-library-is-updated-to-a-new-major-version-with-breaking-changes-how-should-i-update-its-type-declaration-package). You can look at types/node or types/react-dom for usage in practice.

Eggbertx commented:
I've moved sphere1-api.d.ts and sphere2-api.d.ts to v1 and v2 directories, but since the index.d.ts (which previously linked to the APIs) only has comments now, test-all gives a no-useless-files error. Is it acceptable to remove that?

andrewbranch commented:
I think the correct structure would be to put the v2 typings in the root (`types/sphere-engine-browser/index.d.ts`) and v1 in a folder (`types/sphere-engine-browser/v1/index.d.ts`). Then we just have to make sure that the version can be correctly parsed from the header comment of each so they get versioned correctly on npm.

andrewbranch commented on types/sphere-engine-browser/index.d.ts:
```diff
@@ -0,0 +1,3001 @@
+// Type definitions for the non-npm package sphere-engine-browser 5.5
```
This will be published as `@types/sphere-engine-browser@5.5.X`. What is this number? You previously described this as the “v2 API,” so I was expecting this number to start with 2.

Eggbertx replied:
5.5 is the current miniSphere version, I figured it would make sense to have the package follow the engine version when changes are made to the API. And it's referred to as the "v2" API in an abstract sense. There had been different attempts at making an official engine update, since the last official Sphere release was v1.5, with an unfinished 1.6 beta (using the v1.x API)

peterblazejewicz commented on types/sphere-engine-browser/sphere-engine-browser-tests.ts:
```diff
@@ -0,0 +1,71 @@
+/*
+ *  <title of your new Sphere game goes here>
+ *  (c) <year> <whoever made the game>
+ */
+
```
```suggestion
```

fatcerberus commented:
To clarify further on versioning in case it helps guide decisions: v1 and v2 refer to the API revision *only* and are unrelated to the version of miniSphere or Oozaru. This is by design, to ensure API stability. miniSphere and Oozaru are merely implementations and are versioned separately. This situation is similar to ECMAScript revision year vs. browser version. It just so happens that, currently, we have the only viable implementations. :wink:

The API itself *is* versioned, but perhaps not the way one would expect. The **API version** (`Sphere.Version`) is fixed at 2 for the foreseeable future; to bump this would indicate some insanely major breaking change and would be done basically in emergency situations only. Otherwise, new additions to the API (almost always backward compatible) are field-tested in an experimental phase, and then the "API level" (`Sphere.APILevel`) is bumped whenever new functions are finalized.  That value currently stands at 3.

Eggbertx commented:
Alright, I removed the comments from the tests file and changed the API to match the API as reported by the miniSphere engine, let me know if there's anything else I need to change/fix.

andrewbranch reviewed (approved):
I think this looks good. We’ll keep an eye on the publish process and make sure the results look how you want. Thanks all!

Eggbertx commented:
Ready to merge

Eggbertx commented:
I just realized I had overlooked part of the Sphere API used for its build system, since those functions and classes weren't included in the original declaration file. When I add those, will I need to change the package version, or am I able to keep it at 2.0.0 since the API itself hasn't changed?

andrewbranch commented:
The way the versioning works is the major/minor is whatever you put in the header comment, and the patch increments with every publish. So if you make another PR fixing stuff up, it will publish as 2.0.1.
